### PR TITLE
Add rbac support

### PR DESCRIPTION
**Pull request by yxxhero**
Signed-off-by: yxxhero <aiopsclub@163.com>

## Description
add rbac support

## Related Issue
https://github.com/dragonflyoss/Dragonfly2/issues/376

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist:

- [x] Code compiles correctly.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: manager/database/database.go [hunk 1]
```diff
@@ -72,6 +78,29 @@ func migrate(db *gorm.DB) error {
 	)
 }
 
+func initAdminUser(db *gorm.DB) error {
```
Comment (gaius-qi) [inline manager/database/database.go:81]: Put it in the `seed` function.

File: manager/database/database.go [hunk 2]
```diff
@@ -72,6 +78,29 @@ func migrate(db *gorm.DB) error {
...
+	adminUserName := "admin"
+	user := model.User{}
+	err := db.Where("name = ?", adminUserName).First(&user).Error
```
Comment (gaius-qi) [inline manager/database/database.go:84]: Use `Count()` api.

File: manager/handlers/rbac.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+// @Summary Destroy Permission
+// @Description Destroy Permission by json config
+// @Tags permission
+// @Accept json
+// @Produce json
+// @Success 200
+// @Failure 400 {object} HTTPError
+// @Failure 500 {object} HTTPError
+// @Router /permission [post]
```
Comment (gaius-qi) [inline manager/handlers/rbac.go:61]: Use `delete` action.

File: manager/service/service.go [hunk 4]
```diff
@@ -92,3 +99,12 @@ func NewREST(options ...Option) REST {
 
 	return s
 }
+
+//GetRestDB returns the gorm database client
+func GetRestDB(r REST) (*gorm.DB, error) {
```
Comment (gaius-qi) [inline manager/service/service.go:104]: Remove `GetRestDB`.

File: manager/middlewares/rbac.go [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+import (
+	"d7y.io/dragonfly/v2/manager/service"
+	"d7y.io/dragonfly/v2/pkg/permission/rbac"
+	"github.com/casbin/casbin"
+	"github.com/gin-gonic/gin"
+)
+
+func RBAC(r service.REST) (gin.HandlerFunc, *casbin.Enforcer, error) {
+	gdb, err := service.GetRestDB(r)
```
Comment (gaius-qi) [inline manager/middlewares/rbac.go:11]: Get database from `database` package.

File: manager/server/router.go [hunk 6]
```diff
@@ -73,8 +79,14 @@ func initRouter(verbose bool, service service.REST) (*gin.Engine, error) {
...
 	ai.GET("/logout", jwt.LogoutHandler)
-	ai.POST("/refresh_token", jwt.RefreshHandler)
-	ai.POST("/register", h.Register)
+	ai.POST("/refresh_token", jwt.MiddlewareFunc(), jwt.RefreshHandler)
+	ai.POST("/register", jwt.MiddlewareFunc(), rbac, h.Register)
+
+	// Permission
+	pn := apiv1.Group("/permission", jwt.MiddlewareFunc(), rbac)
+	pn.POST("", h.CreatePermission)
+	pn.DELETE("", h.DestroyPermission)
```
Comment (gaius-qi) [inline manager/server/router.go:88]: Delete permission with `ID`.

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: pkg/permission/rbac/rbac.go [hunk 7]
```diff
@@ -0,0 +1,39 @@
+package rbac
```
Comment (gaius-qi) [inline pkg/permission/rbac/rbac.go:1]: Move rbac package to manager.

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: manager/middlewares/rbac.go [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+	if err != nil {
+		return nil, nil, err
+	}
+	enforcer, err := rbac.NewEnforcer(gdb)
```
Comment (gaius-qi) [inline manager/middlewares/rbac.go:15]: Move `NewEnforcer` to `server.New` func.

File: manager/middlewares/rbac.go [hunk 9]
```diff
@@ -0,0 +1,37 @@
+package middlewares
+
+import (
+	"d7y.io/dragonfly/v2/manager/service"
+	"d7y.io/dragonfly/v2/pkg/permission/rbac"
+	"github.com/casbin/casbin"
+	"github.com/gin-gonic/gin"
+)
+
+func RBAC(r service.REST) (gin.HandlerFunc, *casbin.Enforcer, error) {
```
Comment (gaius-qi) [inline manager/middlewares/rbac.go:10]: `func RBAC(e *casbin.Enforcer) gin.HandlerFunc {}`

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: manager/server/router.go [hunk 10]
```diff
@@ -8,14 +8,14 @@ import (
 	ginprometheus "github.com/mcuadros/go-gin-prometheus"
 )
 
-func initRouter(verbose bool, service service.REST) (*gin.Engine, error) {
+func initRouter(verbose bool, serviceREST service.REST) (*gin.Engine, error) {
```
Comment (gaius-qi) [inline manager/server/router.go:11]: Use `servive`. `initRouter` is only used for `rest api`.

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: manager/server/router.go [hunk 11]
```diff
@@ -30,6 +30,12 @@ func initRouter(verbose bool, service service.REST) (*gin.Engine, error) {
 	if err != nil {
 		return nil, err
 	}
+	rbac, enforcer, err := middlewares.RBAC(serviceREST)
```
Comment (gaius-qi) [inline manager/server/router.go:33]: Use middleware like https://github.com/gin-gonic/gin#using-middleware

**Review by gaius-qi — CHANGES_REQUESTED** [review]

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: manager/handlers/rbac.go [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+}
+
+// @Summary Get User Roles
+// @Description Get User Roles
+// @Tags permission
+// @Produce json
+// @Success 200 {object} RoutesInfo
+// @Failure 400 {object} HTTPError
+// @Failure 500 {object} HTTPError
+// @Router /permission/roles/{userName} [get]
```
Comment (gaius-qi) [inline manager/handlers/rbac.go:35]: `/permission/roles/{subject}`

File: manager/middlewares/rbac.go [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+			c.Next()
+			return
+		}
+		// request method
+		m := c.Request.Method
+		action := rbac.HTTPMethodToAction(m)
+		// rbac validation
+		res, err := e.Enforce(userName, permissionGroupName, action)
+		if err != nil || !res {
+			c.JSON(401, gin.H{
```
Comment (gaius-qi) [inline manager/middlewares/rbac.go:25]: Use `http.StatusUnauthorized `

File: manager/server/router.go [hunk 14]
```diff
@@ -88,5 +99,12 @@ func initRouter(verbose bool, service service.REST) (*gin.Engine, error) {
 
 	// Health Check
 	r.GET("/healthy/*action", h.GetHealth)
+
+	// auto init role check roles
```
Comment (gaius-qi) [inline manager/server/router.go:103]: `Auto`

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: manager/handlers/rbac.go [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+	if err := ctx.ShouldBindJSON(&json); err != nil {
+		ctx.JSON(http.StatusUnprocessableEntity, gin.H{"errors": err.Error()})
+		return
+	}
+	err := h.Service.CreatePermission(json)
+	if err != nil {
+		ctx.Error(err)
+		return
+	}
+	ctx.JSON(http.StatusOK, gin.H{"message": "create permission successfully"})
```
Comment (gaius-qi) [inline manager/handlers/rbac.go:100]: Remove message.

File: manager/handlers/rbac.go [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+	if err := ctx.ShouldBindJSON(&json); err != nil {
+		ctx.JSON(http.StatusUnprocessableEntity, gin.H{"errors": err.Error()})
+		return
+	}
+	err := h.Service.DestroyPermission(json)
+	if err != nil {
+		ctx.Error(err)
+		return
+	}
+	ctx.JSON(http.StatusOK, gin.H{"message": "destroy permission successfully"})
```
Comment (gaius-qi) [inline manager/handlers/rbac.go:124]: Remove message.

**Review by gaius-qi — CHANGES_REQUESTED** [review]

**Review by gaius-qi — APPROVED** [review]
LGTM

**Review by gaius-qi — APPROVED** [review]
LGTM

**Review by jim3ma — APPROVED** [review]

**Pull request merged**